### ISSUE TITLE
Two Joins with Alias along the same Tables => Broken Query

### ISSUE THREAD

**Issue by yamass** [jOOQ/jOOQ#13517]
### Expected behavior

When joining like this: 

``` 
       AA
       / \
    BB     BB
    /        \
 CC           CC
```
using aliases and `onKey(...)`, the queries should honor the aliases correctly.

In other words, this test should pass:

```java
/*

CREATE TABLE AA
(
    ID SERIAL PRIMARY KEY
);

CREATE TABLE BB
(
    ID   SERIAL PRIMARY KEY,
    AA_ID INTEGER NOT NULL REFERENCES AA
);

CREATE TABLE CC
(
    ID   SERIAL PRIMARY KEY,
    BB_ID INTEGER NOT NULL REFERENCES BB
);

 */

class JoinAliasTest {

	private static final @NotNull DSLContext DSL_CONTEXT = DSL.using(SQLDialect.POSTGRES, new Settings().withRenderFormatted(true).withRenderSchema(false));

	private static final Bb BB1 = BB.as("bb1");
	private static final Cc CC1 = CC.as("cc1");
	private static final Bb BB2 = BB.as("bb2");
	private static final Cc CC2 = CC.as("cc2");


	@Test
	void test() {
		String incorrectQuery = DSL_CONTEXT.select()
				.from(AA)

				.leftJoin(BB1).onKey(BB1.AA_ID)
				.leftJoin(CC1).onKey(CC1.BB_ID)

				.leftJoin(BB2).onKey(BB2.AA_ID)
				.leftJoin(CC2).onKey(CC2.BB_ID)
				.getSQL(ParamType.INLINED);

		String correctQuery = DSL_CONTEXT.select()
				.from(AA)

				.leftJoin(BB1).on(BB1.AA_ID.eq(AA.ID))
				.leftJoin(CC1).on(CC1.BB_ID.eq(BB1.ID))

				.leftJoin(BB2).on(BB2.AA_ID.eq(AA.ID))
				.leftJoin(CC2).on(CC2.BB_ID.eq(BB2.ID))
				.getSQL(ParamType.INLINED);

		Assertions.assertThat(incorrectQuery)
				.isEqualTo(correctQuery);
	}
}
```

### Actual behavior

We get some weird query that is wrong in many ways...

```sql
select
    "aa"."id",
    "bb1"."id",
    "bb1"."aa_id",
    "cc1"."id",
    "cc1"."bb_id",
    "bb2"."id",
    "bb2"."aa_id",
    "cc2"."id",
    "cc2"."bb_id"
  from "aa"
  left outer join "bb" as "bb1"
    on "bb1"."aa_id" = "aa"."id"
  left outer join "cc" as "cc1"
    on "cc1"."bb_id" = "aa"."id" -- cc1.bb_id references BB, not CC
  left outer join "bb" as "bb2"
    on "bb1"."aa_id" = "bb2"."id" -- bb.aa_id references AA, not BB
  left outer join "cc" as "cc2"
    on "cc1"."bb_id" = "cc2"."id" -- cc.bb_id references BB, not CC
```

### Steps to reproduce the problem

Execute the unit test. Should be self-explanatory.

### Versions

- jOOQ: 3.16.6
- Java: 17.0.2 // locally only ;-) 17.0.3 on server.
- Database (include vendor): Postgres
- OS: MacOS/Linux
- JDBC Driver (include name if inofficial driver): does not matter for this bug.

**Comment by lukaseder** [comment]
Thanks for your report. Probably a duplicate of this? https://github.com/jOOQ/jOOQ/issues/12214. The other report is using `ForeignKey` references, which can't resolve the ambiguity. In your case, it might be possible.

**Comment by yamass** [comment]
@lukaseder You mean like this?

```
		String incorrectQuery = DSL_CONTEXT.select()
				.from(AA)

				.leftJoin(BB1).onKey(Keys.BB__BB_AA_ID_FKEY)
				.leftJoin(CC1).onKey(Keys.CC__CC_BB_ID_FKEY)

				.leftJoin(BB2).onKey(Keys.BB__BB_AA_ID_FKEY)
				.leftJoin(CC2).onKey(Keys.CC__CC_BB_ID_FKEY)
				.getSQL(ParamType.INLINED);
```

How would this solve the problem? That way, the join to CC2 does not know which other join to base on.

Trying this yields the exact same result as the initial query.

By the way, the queries do not only look strange regarding their usage of aliases. The whole structure is messed up. I added corresponding comments to the result in the description.

**Comment by lukaseder** [comment]
I didn't say it would solve the problem. I meant it is even worse.

**Comment by yamass** [comment]
Oh sorry, I somehow misread your comment.

**Comment by lukaseder** [comment]
Finally looking into this now. There's a lot of ambiguity going on in the resolution of `ON KEY` in cases where the same table appears more than once in a join tree. Two things can be stated:

- There are bugs, yes
- There seems to be also a wrong expectation of such paths being resolvable

In the latter case, let's look at your query:

```java
    .from(AA)

    .leftJoin(BB1).onKey(BB1.AA_ID)
    .leftJoin(CC1).onKey(CC1.BB_ID)

    .leftJoin(BB2).onKey(BB2.AA_ID)
    .leftJoin(CC2).onKey(CC2.BB_ID)
```

It is impossible from this query description to decide whether CC2 should be joined to BB2 or BB1. This is more easily illustrated with auxiliary parentheses:

```java
    .from((AA

            .leftJoin(BB1).onKey(BB1.AA_ID)
            .leftJoin(CC1).onKey(CC1.BB_ID)

            .leftJoin(BB2).onKey(BB2.AA_ID))
        .leftJoin(CC2).onKey(CC2.BB_ID) // CC2 is joined to the previous *tree*, not just to BB2
	)
```

It would be different if you had written this, instead:


```java
    .from((AA

            .leftJoin(BB1
                .leftJoin(CC1).onKey(CC1.BB_ID)
            ).onKey(BB1.AA_ID)

            .leftJoin(BB2
			     .leftJoin(CC2).onKey(CC2.BB_ID)
		    ).onKey(BB2.AA_ID))
        
	)
```

Now, there are no more ambiguities. Subtle, hm? This mainly shows that the feature is very brittle and even if the bug parts of this issue are fixed (which I'll do soon), the `ON KEY` syntax should be used with caution. Note that implicit joins are superior to this, because the *exact* join path is always encoded in the syntax:
https://www.jooq.org/doc/latest/manual/sql-building/sql-statements/select-statement/implicit-join/

There is a temptation to "fix" this particular case via convention, saying: The "last" matching table in the join tree should be considered. But what makes the last more preferrable to the "first" (or any "middle" match)? Perhaps we should throw an exception about the ambiguity instead? I'm not sure yet about this. I'm also tempted to deprecate the feature itself, given its unreliability when trees have ambiguities, but I'm aware that many users must have grown fond of it. Perhaps once we support implicit joins / path expressions for to-many joins, the `ON KEY` syntax can be deprecated? https://github.com/jOOQ/jOOQ/issues/13639. This will look more compelling (hypothetical syntax):

```java
    .from(AA)

    .leftJoinImplicit(AA.bb().as("bb1"))
    .leftJoinImplicit(AA.bb().as("bb1").cc().as("cc1"))

    .leftJoinImplicit(AA.bb().as("bb2"))
    .leftJoinImplicit(AA.bb().as("bb2").cc().as("cc2"))
```

Or, using local variables:

```java
BB bb1 = AA.bb().as("bb1");
BB bb2 = AA.bb().as("bb2");
CC cc1 = bb1.cc().as("cc1");
CC cc2 = bb1.cc().as("cc2");


    .from(AA)

    .leftJoinImplicit(bb1)
    .leftJoinImplicit(cc1)

    .leftJoinImplicit(bb2)
    .leftJoinImplicit(cc2)
```

**Comment by lukaseder** [comment]
Looking into the bugs part of this issue, there's already a simpler query that produces wrong results. If I remove the two `CC` tables from the tree, I still get a wrong `JOIN` expression. I'll fix that first: https://github.com/jOOQ/jOOQ/issues/14668

**Comment by lukaseder** [comment]
I can confirm that when nesting joins now with #14668 fixed, then the query is correct, see this comment here: https://github.com/jOOQ/jOOQ/issues/14668#issuecomment-1439796068

**Comment by lukaseder** [comment]
I'm closing this as a duplicate of https://github.com/jOOQ/jOOQ/issues/14668:

- The bug that I've confirmed from your description was fixed with #14668 
- The expectation that `CC2` can be joined correctly is wrong. However, throwing an exception would be backwards incompatible. Ambiguous match warnings are logged already

**Comment by lukaseder** [comment]
Related issue https://github.com/jOOQ/jOOQ/issues/12214 is still open
